File: app/lib/features/reviews/data/reviews_repository.dart
import 'package:sbc_contacts/core/network/api_client.dart';
import 'package:sbc_contacts/features/reviews/domain/review.dart';

class ReviewsRepository {
  ReviewsRepository(this._api);
  final ApiClient _api;

  /// A page of reviews for a member, with the aggregate score and the caller's
  /// own review (backend `MemberReviewsResult`).
  Future<MemberReviews> list(String memberSbcId, {int page = 1, int limit = 20}) async {
    final data = await _api.get('/reviews/$memberSbcId', query: {'page': page, 'limit': limit})
        as Map<String, dynamic>;
    final items = (data['items'] as List<dynamic>? ?? <dynamic>[])
        .whereType<Map<String, dynamic>>()
        .map(Review.fromJson)
        .toList();
    final my = data['myReview'];
    return MemberReviews(
      reviews: items,
      summary: ScoreSummary.fromJson(data['summary'] as Map<String, dynamic>? ?? const {}),
      hasMore: data['hasMore'] as bool? ?? false,
      page: (data['page'] as num?)?.toInt() ?? page,
      myReview: my is Map<String, dynamic> ? Review.fromJson(my) : null,
    );
  }

  /// Create or update the caller's review (upsert). Returns the fresh aggregate.
  Future<ScoreSummary> submit(String memberSbcId, {required int stars, String? comment}) async {
    final data = await _api.post('/reviews', body: {
      'memberSbcId': memberSbcId,
      'stars': stars,
      if (comment != null && comment.isNotEmpty) 'comment': comment,
    }) as Map<String, dynamic>;
    return ScoreSummary.fromJson(data['summary'] as Map<String, dynamic>? ?? const {});
  }

  /// Delete the caller's own review. Returns the fresh aggregate.
  Future<ScoreSummary> remove(String memberSbcId) async {
    final data = await _api.delete('/reviews/$memberSbcId') as Map<String, dynamic>;
    return ScoreSummary.fromJson(data);
  }
}

File: ios/SBCContacts/Features/Reviews/ReviewsRepository.swift
import Foundation

struct ReviewsRepository: Sendable {
    let api: APIClient

    private struct SubmitBody: Encodable, Sendable {
        let memberSbcId: String
        let stars: Int
        let comment: String?
    }

    private struct SubmitResult: Decodable, Sendable {
        let summary: ScoreSummary

        private enum CodingKeys: String, CodingKey { case summary }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            summary = (try? c.decodeIfPresent(ScoreSummary.self, forKey: .summary)) ?? ScoreSummary()
        }
    }

    /// A page of reviews for a member, with the aggregate score and the caller's
    /// own review.
    func list(memberSbcId: String, page: Int = 1, limit: Int = 20) async throws -> MemberReviews {
        try await api.get(
            "/reviews/\(memberSbcId)",
            query: [.init(name: "page", value: String(page)), .init(name: "limit", value: String(limit))]
        )
    }

    /// Create or update the caller's review (upsert). Returns the fresh aggregate.
    @discardableResult
    func submit(memberSbcId: String, stars: Int, comment: String?) async throws -> ScoreSummary {
        let trimmed = comment?.isEmpty == false ? comment : nil
        let result: SubmitResult = try await api.post(
            "/reviews",
            body: SubmitBody(memberSbcId: memberSbcId, stars: stars, comment: trimmed)
        )
        return result.summary
    }

    /// Delete the caller's own review. Returns the fresh aggregate.
    @discardableResult
    func remove(memberSbcId: String) async throws -> ScoreSummary {
        try await api.delete("/reviews/\(memberSbcId)")
    }
}
